### PR TITLE
Relay refactor

### PR DESCRIPTION
**Pull request by jcorbin**
- continues the `foo.extendLogInfo(info)` pattern towards more consistent logging
- pulls key parts out of RelayRequest, to be driven by RelayHandler, to survive in the upcoming lazy path
- drops branched closure allocations
- moves closure allocations to constructor, rather than maybe called methods

r @kriskowal 

cc @jwolski: @Raynos tells me that Ringpop's been using this API, and that this change will require an update in its forwarding path.  We don't consider this a public API nor a breaking change, so major won't be bumping... however I will give you a minor bump, it won't be a 2.6.x.  Note also the crucial change from your perspective is https://github.com/uber/tchannel/commit/c8be0373216a6e0813f49f2d42c1b54a0ad33fb9, which I could ultimately leave out of this branch, and punt the peer selection work until a later lazy branch, let me know how you want to handle this.

### CODE REVIEW

File: node/in_request.js [hunk 1]
```diff
@@ -69,6 +69,24 @@ inherits(TChannelInRequest, EventEmitter);
...
+TChannelInRequest.prototype.extendLogInfo = function extendLogInfo(info) {
+    var self = this;
+
+    // TODO: add:
+    // - request id?
+    // - tracing id?
+    // - other?
+    info.inRemoteAddr = self.remoteAddr;
+    info.serviceName = self.serviceName;
+    if (self.endpoint !== null) {
```
Comment (kriskowal) [inline node/in_request.js]: Perhaps `info.arg1 = null` up front to improve probability of producing a consistent hidden class.

**Comment by kriskowal** [comment]
:ok:

**Comment by jwolski** [comment]
@jcorbin Go ahead and do your thing (merged as is). I'll freeze the TChannel dependency in Ringpop and make sure to update the Ringpop code that depends on this API when we bump the TChannel minor.

**Comment by jcorbin** [comment]
@wolski this will ship in 2.7.0

**Comment by jwolski** [comment]
noted

**Pull request merged**